capnp/src/lib.rs

```diff
@@ -257,21 +473,23 @@ impl core::convert::From<alloc::str::Utf8Error> for Error {
 
 impl core::convert::From<NotInSchema> for Error {
     fn from(e: NotInSchema) -> Self {
-        Self::failed(format!(
-            "Enum value or union discriminant {} was not present in schema.",
-            e.0
-        ))
+        Self::from_kind(ErrorKind::EnumValueOrUnionDiscriminantNotPresent(e))
     }
 }
 
 impl core::fmt::Display for Error {
     fn fmt(&self, fmt: &mut core::fmt::Formatter) -> core::result::Result<(), core::fmt::Error> {
-        write!(fmt, "{:?}: {}", self.kind, self.description)
+        #[cfg(feature = "alloc")]
+        let result = write!(fmt, "{:?}: {}", self.kind, self.description);
+        #[cfg(not(feature = "alloc"))]
+        let result = write!(fmt, "{:?}", self.kind);
```

Comment:
(Note that you write these messages just fine in an `impl Display` without needing `alloc`.)

Reply:
I needed a `#[cfg(feature = "alloc")]` guard because the field `description` is under a `#[cfg(feature = "alloc")]` guard. One way to do this would be to get rid entirely of the `description` String and using the enum error variants everywhere (in `capnp-rpc`, `capnp-futures` etc.). I opted to replace the String with enum only in `capnp` to avoid making too many changes. What do you think, should this be done everywhere?